### PR TITLE
Controller

### PR DESCRIPTION
**Pull request by HatemMorgan**

### CODE REVIEW

File: src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import com.linkedin.replica.serachEngine.config.Configuration;
+
+public class ControllerCommand extends Command {
+	public ControllerCommand(HashMap<String, Object> args) {
+		super(args);
+	}
+
+	private static Configuration config = Configuration.getInstance();
+
+
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java]: Could you please remove the extra blank lines here?

File: src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	private static void deleteFile(String fileName, String packageName) throws IOException{
+		// get .class folder path from configuration
+		String folderPath = config.getAppConfigProp("app.classes.path");
+		// get full path of new file
+		Path path = Paths.get(folderPath + "/" + packageName + "/" + fileName+".class");
+		// delete file if exist to avoid throwing FileAlreadyExistsException  
+		Files.deleteIfExists(path);
+	}
+	
+	public static void main(String[] args) throws IOException {
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java]: Do we need the main method here? Maybe we can move its implementation to some unittest?

File: src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		JsonObject obj = convert(val);
+		
+		// validate embedded JSON object obj to check that all obligatory key value pairs are available
+		validate(obj, new String[]{"packageName","fileName", "configPropKey", "bytes"});
+		
+		// remove double quotations of string value in JSON object
+		String s = obj.get("bytes").toString().replaceAll("\"", "");
+		String packageName = obj.get("packageName").toString().replaceAll("\"", "");
+		String fileName = obj.get("fileName").toString().replaceAll("\"", "");
+//		String commandConfigPropKey = obj.get("configPropKey").toString().replaceAll("\"", "");
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/commands/impl/ControllerCommand.java]: Can you remove this comment?

File: src/main/java/com/linkedin/replica/serachEngine/config/Configuration.java [hunk 4]
```diff
@@ -47,11 +68,70 @@ public Class getHandlerClass(String commandName) throws ClassNotFoundException {
...
+    		writeConfig(commandsConfigPath, commandConfig);
+    		isCommandsConfigModified = false;
+    	}
+    }
+    
+    private void writeConfig(String filePath, Properties properties) throws IOException{
+    	// delete configuration file and then re-write it 
+    	Files.deleteIfExists(Paths.get(filePath));
+    	OutputStream out = new FileOutputStream(filePath);
+    	properties.store(out, "");
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/config/Configuration.java:139]: Please close the `OutputStream` here

File: src/main/java/com/linkedin/replica/serachEngine/controller/Server.java [hunk 5]
```diff
@@ -83,6 +85,13 @@ public void operationComplete(ChannelFuture ch) throws Exception {
 			workerGroup.shutdownGracefully();
 			bossGroup.shutdownGracefully();
 		}
-        
+    
+	}
+	
+	public static void main(String[] args) throws InterruptedException, IOException {
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/controller/Server.java]: Do we need this main method? we should start the controller from the project main right? Also, the host and port should be read from the config file.

File: src/main/java/com/linkedin/replica/serachEngine/models/ResponseType.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.linkedin.replica.serachEngine.models;
+
+/**
+ * Type of responses. 
+ */
+public enum ResponseType {
+	ErrorResponse, SuccessfulResponse
```
Comment (mostafa-abdullah) [inline src/main/java/com/linkedin/replica/serachEngine/models/ResponseType.java]: What about categorizing this according to the status code? For example `BadRequest`, `ServerError`, `Success`, ..etc. ? What do you think?

**Review by mostafa-abdullah — COMMENTED** [review]
Sho3'l 3aly :muscle:

File: src/main/java/com/linkedin/replica/serachEngine/controller/Server.java (hunk 5 above)
Comment (HatemMorgan) [reply src/main/java/com/linkedin/replica/serachEngine/controller/Server.java]: I was testing the controller netty server only. i will remove it

**Review by mostafa-abdullah — APPROVED** [review]

**Pull request merged**